server/server.go

```diff
@@ -168,31 +178,136 @@ func (s *Server) loadSandbox(id string) error {
+			logrus.Warnf("error parsing metadata for %s: %v, ignoring", container.ID, err2)
 			continue
 		}
-		if err = s.loadSandbox(v.Name()); err != nil {
-			logrus.Warnf("could not restore sandbox %s: %v", v.Name(), err)
+		if metadata.Pod {
+			pods[container.ID] = &metadata
+		} else {
+			podContainers[container.ID] = &metadata
 		}
 	}
-	containerDir, err := ioutil.ReadDir(s.runtime.ContainerDir())
-	if err != nil && !os.IsNotExist(err) {
-		logrus.Warnf("could not read container directory %s: %v", s.runtime.ContainerDir(), err)
+	for containerID, metadata := range pods {
+		if err = s.loadSandbox(containerID); err != nil {
+			logrus.Warnf("could not restore sandbox %s container %s: %v", metadata.PodID, containerID, err)
+		}
 	}
-	for _, v := range containerDir {
-		if !v.IsDir() {
-			continue
+	for containerID := range podContainers {
+		if err := s.loadContainer(containerID); err != nil {
+			logrus.Warnf("could not restore container %s: %v", containerID, err)
 		}
-		if err := s.loadContainer(v.Name()); err != nil {
-			logrus.Warnf("could not restore container %s: %v", v.Name(), err)
+	}
+}
+
+// Update makes changes to the server's state (lists of pods and containers) to
+// reflect the list of pods and containers that are stored on disk, possibly
+// having been modified by other parties
+func (s *Server) Update() {
+	err := s.update()
+	if err != nil {
+		logrus.Warnf("warning updating sandbox and container information: %v", err)
+	}
+}
 
+func (s *Server) update() error {
+	containers, err := s.store.Containers()
+	if err != nil && !os.IsNotExist(err) {
+		logrus.Warnf("could not read containers and sandboxes: %v", err)
+	}
+	newPods := map[string]*storage.RuntimeContainerMetadata{}
+	oldPods := map[string]string{}
+	deadPods := map[string]string{}
+	newPodContainers := map[string]*storage.RuntimeContainerMetadata{}
+	oldPodContainers := map[string]string{}
+	deadPodContainers := map[string]string{}
+	for _, container := range containers {
+		if s.hasSandbox(container.ID) {
+			// FIXME: do we need to reload any info about the sandbox?
+			oldPods[container.ID] = container.ID
+			continue
+		}
+		if s.getContainer(container.ID) != nil {
+			// FIXME: do we need to reload any info about the container?
+			oldPodContainers[container.ID] = container.ID
+			continue
+		}
+		// not previously known, so figure out what it is
+		metadata, err2 := s.storage.GetContainerMetadata(container.ID)
+		if err2 != nil {
+			logrus.Warnf("error parsing metadata for %s: %v, ignoring", container.ID, err2)
+			continue
+		}
+		if metadata.Pod {
+			newPods[container.ID] = &metadata
+		} else {
+			newPodContainers[container.ID] = &metadata
+		}
+	}
+	s.ctrIDIndex.Iterate(func(id string) {
+		if _, ok := oldPodContainers[id]; !ok {
+			// this container's ID wasn't in the updated list -> dead
+			deadPodContainers[id] = id
+		}
+	})
+	for deadPodContainer := range deadPodContainers {
+		// forget this container
+		c := s.getContainer(deadPodContainer)
```

Comment:
Not sure if it's about some files I left on my machine, but I had to add `<nil>` check here to prevent panic on the next line, while trying to access reference that not exists. What conditions has to be met to treat container as dead?